### PR TITLE
Rari merger team timelocks

### PR DESCRIPTION
**Pull request by Joeysantoro**
Adds 4 timelocks:
1 using retro vesting from October 2020
3 using vesting from deployment WITH a DAO clawback

All are quadratic over 5 years

### CODE REVIEW

File: contracts/merger/ExchangerTimelock.sol [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    IERC20 public constant tribe =
+        IERC20(0xc7283b66Eb1EB5FB86327f08e1B5816b0720212B);
+
+    constructor(IExchanger _exchanger, address _timelock) {
+        exchanger = _exchanger;
+        timelock = _timelock;
+    }
+
+    /// @notice exchange RGT to TRIBE and send to timelock
+    function exchangeToTimelock() public {
```
Comment (transmissions11) [inline contracts/merger/ExchangerTimelock.sol:36]: external

File: contracts/merger/ExchangerTimelock.sol [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        uint256 rgtBalance = rgt.balanceOf(address(this));
+
+        rgt.approve(address(exchanger), rgtBalance);
+        exchanger.exchange(rgtBalance);
+
+        assert(rgt.balanceOf(address(this)) == 0);
+
+        uint256 tribeBalance = tribe.balanceOf(address(this));
```
Comment (transmissions11) [inline contracts/merger/ExchangerTimelock.sol:44]: this var isnt used more than once so best to inline in the transfer

File: contracts/merger/ExchangerTimelock.sol [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+        tribe.safeTransfer(timelock, tribeBalance);
+
+        assert(tribe.balanceOf(address(this)) == 0);
+    }
+
+    /// @notice guardian sends back RGT
+    function recoverRGT() public {
```
Comment (transmissions11) [inline contracts/merger/ExchangerTimelock.sol:52]: external?

File: contracts/merger/ExchangerTimelock.sol [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        require(msg.sender == guardian);
+        uint256 rgtBalance = rgt.balanceOf(address(this));
```
Comment (transmissions11) [inline contracts/merger/ExchangerTimelock.sol:54]: this var isnt used more than once so best to inline in the transfer

File: contracts/merger/ExchangerTimelock.sol [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+        tribe.safeTransfer(timelock, tribe.balanceOf(address(this)));
+
+        assert(tribe.balanceOf(address(this)) == 0);
+    }
+
+    /// @notice guardian sends back RGT
+    function recoverRGT() external {
+        require(msg.sender == guardian);
+        rgt.transfer(owner(), rgt.balanceOf(address(this)));
```
Comment (transmissions11) [inline contracts/merger/ExchangerTimelock.sol:52]: you use safe transfer for tribe but not for rgt? imo stick with one for consistency. 

we also dont use safeApprove for rgt, if ur gonna go safe i'd go all out

File: proposals/dao/rariTimelock.ts [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+Rari Exchanger Timelocks
+
+DEPLOY ACTIONS:
+
+1. Deploy QuadraticTimelockedDelegator x 3
+2. Deploy ExchangerTimelock x 3
+
+*/
+
+const beneficiary1 = '0xB8f482539F2d3Ae2C9ea6076894df36D1f632775'; // TODO change to actual Rari addresses
```
Comment (ditchfieldcaleb) [inline proposals/dao/rariTimelock.ts:26]: Is this TODO still a TODO action item?

File: contracts/merger/ExchangerTimelock.sol [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        exchanger = _exchanger;
+        timelock = _timelock;
+    }
+
+    /// @notice exchange RGT to TRIBE and send to timelock
+    function exchangeToTimelock() external {
+        uint256 rgtBalance = rgt.balanceOf(address(this));
+
+        rgt.approve(address(exchanger), rgtBalance);
+        exchanger.exchange(rgtBalance);
```
Comment (ditchfieldcaleb) [inline contracts/merger/ExchangerTimelock.sol:40]: Is this going to have to be specially inserted into the Merkle root computation to have the ability to use the Exchanger?

**Review by ditchfieldcaleb — APPROVED** [review]
Approved pending fixing tests & merge conflict, please see above two comments as well.

File: contracts/merger/ExchangerTimelock.sol [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+ @notice For Rari core contributors to trustlessly maintain incentive alignment
+*/
+contract ExchangerTimelock is Ownable {   
+    using SafeERC20 for IERC20;
+
+    IExchanger public immutable exchanger;
+    address public immutable timelock;
+
+    /// @notice guardian multisig
+    address public constant guardian = 0xB8f482539F2d3Ae2C9ea6076894df36D1f632775;
```
Comment (Joeysantoro) [inline contracts/merger/ExchangerTimelock.sol:23]:
```suggestion
    address public constant guardian = 0x8ace03fc45139fddba944c6a4082b604041d19fc;
```

File: proposals/dao/rariTimelock.ts (hunk 6 above)
Comment (Joeysantoro) [reply proposals/dao/rariTimelock.ts:26]: Still a TODO

File: contracts/merger/ExchangerTimelock.sol (hunk 7 above)
Comment (Joeysantoro) [reply contracts/merger/ExchangerTimelock.sol:40]: No peg exchange doesn't have merkle gate

File: contracts/merger/ExchangerTimelock.sol [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+ @notice For Rari core contributors to trustlessly maintain incentive alignment
+*/
+contract ExchangerTimelock is Ownable {   
+    using SafeERC20 for IERC20;
+
+    IExchanger public immutable exchanger;
+    address public immutable timelock;
+
+    /// @notice guardian multisig
+    address public constant guardian = 0x8ace03fc45139fddba944c6a4082b604041d19fc;
```
Comment (Joeysantoro) [inline contracts/merger/ExchangerTimelock.sol:23]:
```suggestion
    address public constant guardian = 0x8ace03Fc45139fDDba944c6A4082b604041d19FC;
```

File: contract-addresses/mainnetAddresses.ts [hunk 10]
```diff
@@ -20,7 +20,7 @@ const MainnetAddresses = {
   pegExchanger: { artifactName: 'PegExchanger', address: '0xc09BB5ECf865e6f69Fe62A43c27f036A426909f7' },
   rariTimelock: { artifactName: 'Timelock', address: '0x8ace03Fc45139fDDba944c6A4082b604041d19FC' },
   gfxAddress: { artifactName: 'unknown', address: '0xA6E8772AF29B29B9202A073F8E36F447689BEEF6' },
-  rgt: { artifactName: 'IERC20', address: '0xD291E7a03283640FDc51b121aC401383A46cC623' },
+  rgt: { artifactName: 'Tribe', address: '0xD291E7a03283640FDc51b121aC401383A46cC623' },
```
Comment (ditchfieldcaleb) [inline contract-addresses/mainnetAddresses.ts:23]: ??

Comment (ditchfieldcaleb) [reply contract-addresses/mainnetAddresses.ts:23]: Would prefer a rename here

File: proposals/dao/rariTimelock.ts [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+2. Deploy QuadtraticTimelockedSubdelegator
+3. Deploy ExchangerTimelock x2
+
+*/
+
+const delegatorBeneficiary = '0xB8f482539F2d3Ae2C9ea6076894df36D1f632775'; // TODO change to actual Rari addresses
+const subdelegatorBeneficiary = '0xB8f482539F2d3Ae2C9ea6076894df36D1f632775'; // TODO change to actual Rari addresses
+
+const FIVE_YEARS = '157680000';
+const TIMELOCK_START = '1630000000'; // TODO set to Rari vesting start
```
Comment (ElliotFriedman) [inline proposals/dao/rariTimelock.ts:31]: Remove TODO comments and replace with actual values for the DAO script.

File: scripts/utils/exec.ts [hunk 12]
```diff
@@ -28,51 +29,48 @@ async function exec(txData, totalValue, addresses) {
...
   }
 
-  try {
-    await governor.castVote(proposalNo, true, { from: voterAddress });
-    console.log('Casted vote');
-  } catch {
-    console.log('Already voted');
-  }
+  await governor.connect(signer).castVote(proposalNo, 1);
+  console.log('Casted vote');
```
Comment (ElliotFriedman) [inline scripts/utils/exec.ts:53]: remove console log

File: scripts/utils/exec.ts [hunk 13]
```diff
@@ -28,51 +29,48 @@ async function exec(txData, totalValue, addresses) {
...
+  const signer = await getImpersonatedSigner(voterAddress);
 
   console.log(`Proposal Number: ${proposalNo}`);
 
   let proposal = await governor.proposals(proposalNo);
   const { startBlock } = proposal;
 
   // Advance to vote start
-  if ((await time.latestBlock()) < startBlock) {
+  if (toBN(await time.latestBlock()) < toBN(startBlock)) {
```
Comment (ElliotFriedman) [inline scripts/utils/exec.ts:45]: if we're using a BN, I wouldn't think they would work with the `<` symbol as it is an object, not a number.

File: proposals/description/merger.ts [hunk 14]
```diff
@@ -40,7 +40,16 @@ const merger: ProposalDescription = {
     }
   ],
   description: `
-  Code: 
+  This proposal represents the Tribe half of the FeiRari merger code. It executes the following steps:
+  1. Accept PegExchanger contract for swapping RGT to TRIBE at ~26.7 TRIBE per RGT exchange rate
+  2. Accept TRIBERagequit contract
+  3. Seed PegExchanger with 270m TRIBE
+  4. Grant FEI minting to TRIBERagequit
```
Comment (ElliotFriedman) [inline proposals/description/merger.ts:47]: A safer pattern would be to mint FEI to the TRIBERagequit on a dripper rather than giving that contract minting abilities.

File: contracts/merger/ExchangerTimelock.sol (hunk 5 above)
Comment (ElliotFriedman) [reply contracts/merger/ExchangerTimelock.sol:52]: agree on consistency. if we use safeTransfer for one, we should also use it for the other

File: contracts/merger/ExchangerTimelock.sol [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+ @notice For Rari core contributors to trustlessly maintain incentive alignment
+*/
+contract ExchangerTimelock is Ownable {   
+    using SafeERC20 for IERC20;
+
+    IExchanger public immutable exchanger;
+    address public immutable timelock;
+
+    /// @notice guardian multisig
+    address public constant guardian = 0x8ace03Fc45139fDDba944c6A4082b604041d19FC;
```
Comment (ditchfieldcaleb) [inline contracts/merger/ExchangerTimelock.sol:23]: should note that this is the *rari* timelock, not the fei-guardian multisig

File: proposals/dao/rariTimelock.ts [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+
+DEPLOY ACTIONS:
+
+1. Deploy QuadraticTimelockedDelegator
+2. Deploy QuadtraticTimelockedSubdelegator
+3. Deploy ExchangerTimelock x2
+
+*/
+
+const delegatorBeneficiary = '0xB8f482539F2d3Ae2C9ea6076894df36D1f632775'; // TODO change to actual Rari addresses
```
Comment (ditchfieldcaleb) [inline proposals/dao/rariTimelock.ts:27]: fix todos

File: proposals/dao/rariTimelock.ts [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+const FIVE_YEARS = '157680000';
+const TIMELOCK_START = '1630000000'; // TODO set to Rari vesting start
+
+const toBN = ethers.BigNumber.from;
+
+export const deploy: DeployUpgradeFunc = async (deployAddress, addresses, logging = false) => {
+  const { pegExchanger, tribe } = addresses;
+
+  if (!pegExchanger || !tribe) {
+    throw new Error('An environment variable contract address is not set');
```
Comment (ditchfieldcaleb) [inline proposals/dao/rariTimelock.ts:39]: nit: these aren't environment variables

Comment (ditchfieldcaleb) [reply proposals/dao/rariTimelock.ts:39]: (environment variable has a very specific meaning so I'm going to keep saying this)

File: proposals/dao/rariTimelock.ts [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+
+const delegatorBeneficiary = '0xB8f482539F2d3Ae2C9ea6076894df36D1f632775'; // TODO change to actual Rari addresses
+const subdelegatorBeneficiary = '0xB8f482539F2d3Ae2C9ea6076894df36D1f632775'; // TODO change to actual Rari addresses
+
+const FIVE_YEARS = '157680000';
+const TIMELOCK_START = '1630000000'; // TODO set to Rari vesting start
+
+const toBN = ethers.BigNumber.from;
+
+export const deploy: DeployUpgradeFunc = async (deployAddress, addresses, logging = false) => {
```
Comment (ditchfieldcaleb) [inline proposals/dao/rariTimelock.ts:35]: Can you add more comments in this file? It looks generally okay but it's kinda hard to parse out what all is going on and why.

**Review by ditchfieldcaleb — CHANGES_REQUESTED** [review]

File: proposals/description/merger.ts (hunk 14 above)
Comment (ditchfieldcaleb) [reply proposals/description/merger.ts:47]: It's too late to change this; the merger is already up for vote on Rari.

File: proposals/dao/rariTimelock.ts (hunk 11 above)
Comment (Joeysantoro) [reply proposals/dao/rariTimelock.ts:31]: yup need values from @davidlucid

File: scripts/utils/exec.ts (hunk 13 above)
Comment (Joeysantoro) [reply scripts/utils/exec.ts:45]: It appears that this works. They must have comparators built-in. I can change to bn method

File: contracts/merger/ExchangerTimelock.sol (hunk 15 above)
Comment (Joeysantoro) [reply contracts/merger/ExchangerTimelock.sol:23]: changed address not comment, will fix

File: contract-addresses/mainnetAddresses.ts (hunk 10 above)
Comment (Joeysantoro) [reply contract-addresses/mainnetAddresses.ts:23]: Will fix

File: proposals/dao/rariTimelock.ts (hunk 16 above)
Comment (Joeysantoro) [reply proposals/dao/rariTimelock.ts:27]: yes

File: proposals/dao/rariTimelock.ts (hunk 17 above)
Comment (Joeysantoro) [reply proposals/dao/rariTimelock.ts:39]: k will fix

File: proposals/dao/rariTimelock.ts (hunk 18 above)
Comment (Joeysantoro) [reply proposals/dao/rariTimelock.ts:35]: will do

File: scripts/utils/exec.ts (hunk 13 above)
Comment (ElliotFriedman) [reply scripts/utils/exec.ts:45]: You don't have to use the bn built in, you could just cast to a type Number

File: contracts/merger/PegExchangerDripper.sol [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+
+//     }
+ 
+//     /// @notice drip ERC20 tokens to target
+//     function drip()
+//        external
+//     {
+
+//         emit Dripped(amountToDrip);
+//     }
```
Comment (ElliotFriedman) [inline contracts/merger/PegExchangerDripper.sol:24]: if the whole file is commented, delete from git

File: contracts/dao/QuadraticTimelockedSubdelegator.sol [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+    /// @param _startTime the initial time to use for timelock
+    constructor(
+        address _beneficiary,
+        uint256 _duration,
+        address _tribe,
+        uint256 _cliff,
+        address _clawbackAdmin,
+        uint256 _startTime
+    ) QuadraticTokenTimelock(_beneficiary, _duration, _tribe, _cliff, _clawbackAdmin, _startTime) {
+        tribe = ITribe(_tribe);
```
Comment (eswak) [inline contracts/dao/QuadraticTimelockedSubdelegator.sol:62]: Note that if we set this timelock up for someone, they could delegate all of their tokens to another address. The TRIBE gets transferred to a new Delegatee contract, so it is then out of reach from clawback.

Maybe add a require(_clawbackAdmin == address(0)) in the constructor, to make sure this kind of timelock is only used for people that don't have a clawback.

File: contracts/utils/timelock/QuadraticTokenTimelock.sol [hunk 21]
```diff
@@ -8,14 +8,21 @@ contract QuadraticTokenTimelock is TokenTimelock {
...
-        7776000, // 3 months cliff
+        _cliffDuration,
         _lockedToken, 
-        address(0xB8f482539F2d3Ae2C9ea6076894df36D1f632775) // fei labs multisig
-    ) {}
+        _clawbackAdmin
+    ) {
+        if (_startTime != 0) {
+            startTime = _startTime;
+        }
```
Comment (eswak) [inline contracts/utils/timelock/QuadraticTokenTimelock.sol:24]: it should revert like crazy if we set a startTime in the future, because of checks like this in `Timed.sol`  ?
```
uint256 timePassed = block.timestamp - startTime; // block timestamp always >= startTime
```

File: contracts/dao/QuadraticTimelockedSubdelegator.sol (hunk 20 above)
Comment (eswak) [reply contracts/dao/QuadraticTimelockedSubdelegator.sol:62]: The Rari team subdelegator does have a clawbackAdmin, so I think you should override the clawback function here to recall tokens from all delegatees

**Review by eswak — DISMISSED** [review]

File: contracts/dao/QuadraticTimelockedSubdelegator.sol (hunk 20 above)
Comment (Joeysantoro) [reply contracts/dao/QuadraticTimelockedSubdelegator.sol:62]: Fixing this, enforcing clawback admin must be 0

File: contracts/utils/timelock/QuadraticTokenTimelock.sol (hunk 21 above)
Comment (Joeysantoro) [reply contracts/utils/timelock/QuadraticTokenTimelock.sol:24]: This is fine imo. The use case here is actually in the *past* for rari founders who started vesting on October 2020

**Review by eswak — APPROVED** [review]

**Review by eswak — DISMISSED** [review]
